### PR TITLE
Add support for SQL Database SKU and MaxSizeBytes

### LINKED ISSUE #1226: Add support for AzureSQLDatabase SKU and disk size
**Describe the current behavior**
SQLDatabase is supported, but Sku and disk size are not.

**Describe the improvement**
Add ability to specify the SKU and Disk size for Azure SQL Databases. The current implementation of AzureSQLDatabase is not flexible enough to allow user to provision specific SKUs that Azure offers.

### PR DESCRIPTION
**Pull request by matthchr**
Closes #1226

**What this PR does / why we need it**:
Allow customers full control over SQL server SKU and MaxSize

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains tests

### CODE REVIEW

File: controllers/azuresql_combined_test.go [hunk 1]
```diff
@@ -106,7 +109,31 @@ func TestAzureSqlServerCombinedHappyPath(t *testing.T) {
 				},
 			}
 
-			EnsureInstance(ctx, t, tc, sqlDatabaseInstance)
+			EnsureInstance(ctx, t, tc, sqlDatabaseInstance1)
+		})
+
+		t.Run("set up database in primary server using sku with maxsizebytes", func(t *testing.T) {
```
Comment (Porges) [inline controllers/azuresql_combined_test.go:115]: this is the secondary server surely :)

**Comment by Porges** [comment]
Seems to make sense 👍

File: pkg/resourcemanager/azuresql/azuresqlshared/sqlproperties.go [hunk 2]
```diff
@@ -78,48 +96,77 @@ func SQLServerPropertiesToServer(properties SQLServerProperties) (result sql.Ser
...
+	// These defaults were determined based on what SQL defaults to in the 2015-03-01-preview API.
+	// The way this works is that for editions that were DTU based, Edition maps to name: Edition, tier: Edition,
+	// so for example edition: "Basic" became name: "Basic", tier: "Basic". For VCore offerings, the name can't actually
+	// be derived from just the old edition, so we had to hardcode some actual values (the defaults which SQL happens
+	// to use - i.e. HS_Gen5_2 is the default when you deploy a 2015-03-01-preview SQL database with Edition: "Hyperscale").
+	// There is some risk that if SQL changes their defaults we'll be out of sync, but given we're planning to remove
+	// edition in a future API version and customers have full access to SKU to specify exactly what they want, that
+	// shouldn't really be a problem.
 	switch in {
 	case 0:
```
Comment (Porges) [inline pkg/resourcemanager/azuresql/azuresqlshared/sqlproperties.go:149]: can we use named constants here?

File: controllers/azuresql_combined_test.go (hunk 1 above)
Comment (matthchr) [reply controllers/azuresql_combined_test.go:115]: So it is

Comment (matthchr) [reply controllers/azuresql_combined_test.go:115]: Ah sorry, no it's not.

It's the secondary _database_ in the primary _server_ (there are two servers too)

File: pkg/resourcemanager/azuresql/azuresqlshared/sqlproperties.go (hunk 2 above)
Comment (matthchr) [reply pkg/resourcemanager/azuresql/azuresqlshared/sqlproperties.go:149]: Yes -- although the code actually didn't before. The customer just specified it as a number (which seemed weird to me -- not clear where they got the mapping between number and meaning...

**Comment by frodopwns** [comment]
@matthchr because the spec changed here you will need to update the helm templates

```
make helm-chart-manifests
```

then commit the changed files

**Comment by frodopwns** [comment]
Probably should also update the sample manifest and/or the docs https://github.com/Azure/azure-service-operator/blob/master/config/samples/azure_v1beta1_azuresqldatabase.yaml

File: pkg/resourcemanager/azuresql/azuresqldb/azuresqldb.go [hunk 3]
```diff
@@ -108,15 +114,21 @@ func (_ *AzureSqlDbManager) CreateOrUpdateDB(ctx context.Context, resourceGroupN
...
+	}
+
 	return future.Response(), err
 }
 
 // AddLongTermRetention enables / disables long term retention
 func (_ *AzureSqlDbManager) AddLongTermRetention(ctx context.Context, resourceGroupName string, serverName string, databaseName string, weeklyRetention string, monthlyRetention string, yearlyRetention string, weekOfYear int32) (*http.Response, error) {
 
 	longTermClient, err := azuresqlshared.GetBackupLongTermRetentionPoliciesClient()
+	// TODO: Probably shouldn't return a response at all in the err case here (all through this function)
```
Comment (frodopwns) [inline pkg/resourcemanager/azuresql/azuresqldb/azuresqldb.go:131]: I think at times there have been attempts to make the Response more user friendly so we could rely on it for error handling. In the end we have had to just use data embedded in the error types.

File: pkg/resourcemanager/azuresql/azuresqldb/azuresqldb_reconcile.go [hunk 4]
```diff
@@ -45,14 +44,26 @@ func (db *AzureSqlDbManager) Ensure(ctx context.Context, obj runtime.Object, opt
 	if len(instance.Spec.DbName) > 0 {
 		dbName = instance.Spec.DbName
 	}
-	dbEdition := instance.Spec.Edition
+
+	dbSku, err := azuresqlshared.MergeDBEditionAndSku(instance.Spec.Edition, instance.Spec.Sku)
+
```
Comment (frodopwns) [inline pkg/resourcemanager/azuresql/azuresqldb/azuresqldb_reconcile.go:49]:
```suggestion
```

File: pkg/resourcemanager/azuresql/azuresqlmanageduser/azuresqlmanageduser.go [hunk 5]
```diff
@@ -54,7 +54,6 @@ func (s *AzureSqlManagedUserManager) GetDB(ctx context.Context, resourceGroupNam
 		resourceGroupName,
 		serverName,
 		databaseName,
-		"serviceTierAdvisors, transparentDataEncryption",
```
Comment (frodopwns) [inline pkg/resourcemanager/azuresql/azuresqlmanageduser/azuresqlmanageduser.go:57]: mystery solved :-)

Comment (matthchr) [reply pkg/resourcemanager/azuresql/azuresqlmanageduser/azuresqlmanageduser.go:57]: Yeah we dont use it for anything

File: api/v1beta1/azuresqldatabase_types.go [hunk 6]
```diff
@@ -22,13 +49,16 @@ type AzureSqlDatabaseSpec struct {
 	// +kubebuilder:validation:Required
 	ResourceGroup string    `json:"resourceGroup"`
 	Server        string    `json:"server"`
-	Edition       DBEdition `json:"edition"`
+	Edition       DBEdition `json:"edition"` // TODO: Remove this in v1beta2
```
Comment (frodopwns) [inline api/v1beta1/azuresqldatabase_types.go:52]: this now needs to include `omitempty`

**Pull request merged**